pkg/derivation/store/badger.go

```diff
@@ -0,0 +1,200 @@
+package store
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	badger "github.com/dgraph-io/badger/v3"
+	"github.com/nix-community/go-nix/pkg/derivation"
+)
+
+var _ derivation.Store = &BadgerStore{}
+
+// NewBadgerStore opens a store that stores its data
+// in the path specified by path.
+func NewBadgerStore(path string) (*BadgerStore, error) {
+	db, err := badger.Open(badger.DefaultOptions(path).
+		WithLoggingLevel(badger.DEBUG))
+	if err != nil {
+		return nil, fmt.Errorf("error opening badger store: %w", err)
+	}
+
+	return &BadgerStore{
+		db: db,
+	}, nil
+}
+
+// NewBadgerMemoryStore opens a store that entirely resides in memory.
+func NewBadgerMemoryStore() (*BadgerStore, error) {
+	db, err := badger.Open(badger.DefaultOptions("").WithInMemory(true))
+	if err != nil {
+		return nil, fmt.Errorf("error opening badger store: %w", err)
+	}
+
+	return &BadgerStore{
+		db: db,
+	}, nil
+}
+
+// BadgerStore stores data using badger.
+// All derivations are stored in ATerm format, at `drv:$drvPath`.
+// The replacement string for a drv is stored at `replacement:$drvPath`.
+// The interface should be thread-safe.
+type BadgerStore struct {
+	db *badger.DB
+}
+
+// Put inserts a new Derivation into the Derivation Store.
+func (bs *BadgerStore) Put(ctx context.Context, drv *derivation.Derivation) (string, error) {
+	if err := validateDerivationInStore(ctx, drv, bs); err != nil {
```

Comment:
This does trigger requests to `store.Has()` under the hood.

Right now, we don't support removing anything from a `derivation.Store`, but in the future, once we do, we might need to think about this, and potentially expand the transaction to this as well (and then maybe inline the code checking for these input derivations).

If we succeed all checks, and actually insert, we might also need to detect if something else removed any derivation this derivation (transitively) refers to.

Maybe that's also all too complicated, and `Put()` and `Delete()`/`GC()` should just block, I'm not sure - mostly leaving this here for future reference.